controller/src/main/java/io/pravega/controller/store/stream/AbstractInt96Counter.java

```diff
@@ -0,0 +1,127 @@
+/**
+ * Copyright Pravega Authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.pravega.controller.store.stream;
+
+import com.google.common.annotations.VisibleForTesting;
+import io.pravega.common.lang.AtomicInt96;
+import io.pravega.common.lang.Int96;
+import io.pravega.common.tracing.TagLogger;
+import org.slf4j.LoggerFactory;
+
+import javax.annotation.concurrent.GuardedBy;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.CompletionException;
+
+/**
+  * Abstract Int96 counter. It implements getNextCounter using the Int96Counter interface.
+  */
+abstract class AbstractInt96Counter implements Int96Counter {
+    private static final TagLogger log = new TagLogger(LoggerFactory.getLogger(AbstractInt96Counter.class));
+    final Object lock;
+    @GuardedBy("lock")
+    final AtomicInt96 limit;
+    @GuardedBy("lock")
+    final AtomicInt96 counter;
+    @GuardedBy("lock")
+    CompletableFuture<Void> refreshFutureRef;
+
+    public AbstractInt96Counter() {
+        this.lock = new Object();
+        this.counter = new AtomicInt96();
+        this.limit = new AtomicInt96();
+        this.refreshFutureRef = null;
+    }
+
+    @Override
+    public CompletableFuture<Int96> getNextCounter(OperationContext context) {
+        CompletableFuture<Int96> future;
+        synchronized (lock) {
+            Int96 next = counter.incrementAndGet();
+            if (next.compareTo(limit.get()) > 0) {
+                // ignore the counter value and after refreshing call getNextCounter
+                future = refreshRangeIfNeeded(context).thenCompose(x -> getNextCounter(context));
```

Comment:
This is chaining futures inside of a lock. This is both confusing and can lead to inconsistant results in testing because sometimes the code will run with the lock held and other times without it. Please refactor this to make it more clear.